src/Microsoft.VisualBasic/src/Microsoft/VisualBasic/FileIO/SpecialDirectories.vb

```diff
@@ -0,0 +1,244 @@
+        '''     each process has its own copy of the environment variables and this copy is not updated.
+        ''' </remarks>
+        Public Shared ReadOnly Property Temp() As String
+            Get
+                Return GetDirectoryPath(IO.Path.GetTempPath(), SR.IO_SpecialDirectory_Temp)
+            End Get
+        End Property
+
+        ''' <summary>
+        ''' Returns the directory that serves as a common repository for data files
+        ''' from your application used only by the current user.
+        ''' </summary>
+        ''' <value>A String containing the path to the directory your application can use to store data for the current user.</value>
+        ''' <remarks>
+        ''' We chose to use UserAppDataPath instead of LocalUserAppDataPath since this directory
+        ''' will work with Roaming User as well.
+        ''' </remarks>
+        Public Shared ReadOnly Property CurrentUserApplicationData() As String
+            Get
+                Dim ApplicationData As String = GetDirectoryPath(Environment.GetFolderPath(SpecialFolder.ApplicationData, SpecialFolderOption.Create), SR.IO_SpecialDirectory_UserAppData).Trim.TrimEnd(Separators).Trim
+                If ApplicationData.Length = 0 Then
+                    Throw ExUtils.GetDirectoryNotFoundException(SR.IO_SpecialDirectoryNotExist, GetResourceString(SR.IO_SpecialDirectory_UserAppData) & $" SpecialFolder.ApplicationData returned String.Empty.")
+                End If
+                If Not IO.Directory.Exists(ApplicationData) Then
+                    Throw ExUtils.GetDirectoryNotFoundException(SR.IO_SpecialDirectoryNotExist, GetResourceString(SR.IO_SpecialDirectory_UserAppData) & $" SpecialFolder.ApplicationData could not create '{ApplicationData}'")
+                End If
+                Try
+                    Return CreateValidFullPath(ApplicationData)
+                Catch ex As Exception
+                    Throw ExUtils.GetDirectoryNotFoundException(SR.IO_SpecialDirectoryNotExist, GetResourceString(SR.IO_SpecialDirectory_UserAppData) & $" '{ApplicationData}' with real exception {ex.Message}")
+                End Try
+            End Get
+        End Property
+
+        ''' <summary>
+        ''' Returns the directory that serves as a common repository for data files
+        ''' from your application used by all users.
+        ''' </summary>
+        ''' <value>A String containing the path to the directory your application can use to store data for all users.</value>
+        ''' <remarks>
+        ''' If a path does not exist, one is created in the following format
+        ''' C:\Documents and Settings\All Users\Application Data\[CompanyName]\[ProductName]\[ProductVersion]
+        '''
+        ''' See above for reason why we don't use System.Environment.GetFolderPath(*).
+        ''' </remarks>
+        Public Shared ReadOnly Property AllUsersApplicationData() As String
```

Comment:
>AllUsersApplicationData [](start = 40, length = 23)

Should this `Throw New PlatformNotSupportedException()` on non-Windows platforms?

Reply:
See above comment, but I don't think so. Except on Nano I System.IO returns something I should be able to work with. System.IO and what VB expects are very different. The reference source does not use Environment.SpecialDirectories because it only returns the unseparated path. Any help here would be great and the issue in on Windows as well as Linux.

Reply:
There are many changes to `CurrentUserApplicationData` and `AllUsersApplicationData` from the reference source. Is that because `System.Windows.Forms` is not available on .NET Core yet? If so, consider excluding these methods out from this PR so that this PR contains a straight port of the reference source with as few changes as possible.

---
In reply to: [227648484](https://github.com/dotnet/corefx/pull/32668#discussion_r227648484) [](ancestors = 227648484)